src/app/cli/src/commands/delete_command.rs

```diff
@@ -20,57 +21,133 @@ use crate::Interact;
+                            out_of_sync.push((remote, "behind"));
+                        }
+                        CompareChainsResult::Divergence { .. } => {
+                            out_of_sync.push((remote, "diverged from"));
+                        }
+                        _ => (),
+                    },
+                    Err(e) => unknown.push((remote, e)),
+                }
+            }
+
+            let all_synced = out_of_sync.is_empty() && unknown.is_empty();
+
+            if !all_synced {
+                eprintln!(
+                    "{} dataset is out of sync with remote(s):",
+                    console::style(&hdl.alias).bold()
+                );
+                for (remote, status_desc) in out_of_sync {
+                    eprintln!(
+                        " - {} '{}'",
+                        console::style(status_desc).yellow(),
+                        console::style(&remote).bold(),
+                    );
+                }
+                for (remote, err) in unknown {
+                    eprintln!(
+                        " - {} '{}'. Error: {}",
+                        console::style("could not check state of").yellow(),
+                        console::style(&remote).bold(),
+                        err,
+                    );
+                }
+                eprintln!();
+            };
+        }
+
+        self.interact.require_confirmation(format!(
+            "{}\n  {}\n{}",
+            console::style("You are about to delete following dataset(s):").yellow(),
+            itertools::join(dataset_handles.iter().map(|h| &h.alias), "\n  "),
+            console::style("This operation is irreversible!").yellow(),
+        ))?;
+        Ok(())
+    }
 }
 
 #[async_trait::async_trait(?Send)]
 impl Command for DeleteCommand {
     async fn run(&mut self) -> Result<(), CLIError> {
-        if self.dataset_ref_patterns.is_empty() && !self.all {
+        let target_is_empty = match &self.target {
+            DeleteTarget::Handles(hs) => hs.is_empty(),
+            DeleteTarget::Patterns(ps) => ps.is_empty(),
+        };
+
+        if target_is_empty && !self.all {
             return Err(CLIError::usage_error("Specify a dataset or use --all flag"));
         }
 
         let dataset_handles: Vec<DatasetHandle> = if self.all {
             self.dataset_repo.get_all_datasets().try_collect().await?
         } else {
-            filter_datasets_by_local_pattern(
-                self.dataset_repo.as_ref(),
-                self.dataset_ref_patterns.clone(),
-            )
-            .try_collect()
-            .await?
+            match &self.target {
+                DeleteTarget::Handles(hs) => hs.clone(),
+                DeleteTarget::Patterns(ps) => {
+                    filter_datasets_by_local_pattern(self.dataset_repo.as_ref(), ps.clone())
+                        .try_collect()
+                        .await?
+                }
+            }
```

Comment:
Can we avoid clone here? I think you can move instead in both occurrences of `.clone()`